### PR TITLE
Fix #5768, Corrected wrong formatted print of a trait member constraint.

### LINKED ISSUE #5768: SRTP reports a required constraint (twice) that has already been specified
Type resolution reports FS0001 and FS0043 with the same error, that a constraint is required - but that constraint is already specified

#### Repro steps

Crazy code sample:
```fsharp
type Tuple =
    static member inline TupleMap ((a, b),          _ : Tuple) = fun f -> (f a, f b)
    static member inline TupleMap ((a, b, c),       _ : Tuple) = fun f -> (f a, f b, f c)
    static member inline TupleMap ((a, b, c, d),    _ : Tuple) = fun f -> (f a, f b, f c, f d)
    
    static member inline Map f (x: 'Tin) : 'Tout =
        let inline call_2 (a: ^a, b : ^b) = ((^a or ^b) : (static member TupleMap : ^b * _ -> ^t) (b, a))
        call_2 (Unchecked.defaultof<Tuple>, x) f

type IOption<'T> =
    abstract member Value : 'T

module Something = 
    let inline tupleMap f x = Tuple.Map f x

    let inline addOptionValues< ^value, ^options, ^values, 'item when 
                                  'item :> IOption< ^value>
                                  and (Tuple or ^options) : (static member TupleMap : ^options * Tuple -> ('item -> ^value) -> ^values)>
                                  (addUp : ^values -> ^value, sourceOptions : ^options) =
        let getValue (i : 'item) = i.Value
        let allValues : ^values = tupleMap getValue sourceOptions
        let result : ^value = addUp allValues
        result
``` 

The line 
```fsharp
        let allValues : ^values = tupleMap getValue sourceOptions
```
Gives the errors
```
Error	FS0043	A type parameter is missing a constraint 'when (Tuple or  ^options) : (static member TupleMap :  ^options * Tuple -> ('item ->  ^value) ->  ^values)'	
Error	FS0001	A type parameter is missing a constraint 'when (Tuple or  ^options) : (static member TupleMap :  ^options * Tuple -> ('item ->  ^value) ->  ^values)'	
```

#### Related information 

Provide any related information 

* Windows 10
* Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: 55a3dc3231c95c77f81ee53f7d29152029da7408.
* Microsoft Visual F# Tools 10.2 for F# 4.5

### PR DESCRIPTION
**Pull request by realvictorprm**
The bug was not that the constraint is ignored. It was that the compiler printed the suggested trait constraint wrongly. 

As example, the signature `a * b -> (c -> d)` isn't the same as `a * b -> c -> d` because the return type differs (in case one it's `(c -> d)`, in case two it's `d`). The compiler however didn't respect that behaviour and prints constraints of case 1 always like case 2 which is wrong. The fix was to enforce that trait constraint return type is always wrapped in brackets if the return type is a function.

@dsyme please have a look whether this assumption is fine? It's a short thing only about formatting.

Signed-off-by: realvictorprm <mueller.vpr@gmail.com>

### CODE REVIEW

**Comment by realvictorprm** [comment]
CI passes. This is ready.

**Comment by realvictorprm** [comment]
I'll have a look how I can add it.

**Comment by realvictorprm** [comment]
Regression test added.

File: tests/fsharp/tests.fs [hunk 1]
```diff
@@ -1737,6 +1737,22 @@ module RegressionTests =
 
         peverify cfg "test.exe"
 #endif
+#if !FSHARP_SUITE_DRIVES_CORECLR_TESTS
+    [<Test>]
+    let ``5768`` () =
```
Comment (dsyme) [inline tests/fsharp/tests.fs]: I don't understand why these tests aren't a simple new single-file test under `typecheck\sigs\...` like all the other negative tests. You don't have to write all this diffing code manually.

Comment (realvictorprm) [reply tests/fsharp/tests.fs]: I wasn't aware of this! :)
Thanks for the pointer

**Comment by realvictorprm** [comment]
@dsyme I've changed it, please have a short look over it.

**Comment by dsyme** [comment]
approved, please resolve conflict and we can merge

**Comment by realvictorprm** [comment]
done

**Comment by cartermp** [comment]
Ergh, looks like there were some build issues preventing this from getting merged and the codebase has diverged since. @realvictorprm are you still interested in pursuing this? In theory it should be easy to adjust the tests and just copy the contents of this branch's `neg111.fs` and `neg111.bsl` into new files.

**Comment by abelbraaksma** [comment]
@cartermp, in an attempt not to lose this work, I've resurrected it here: https://github.com/dotnet/fsharp/pull/9728 (branched off this one to keep history intact).

**Comment by cartermp** [comment]
Thanks! Will close in favor of https://github.com/dotnet/fsharp/pull/9728

**Pull request closed without merging**